src/mappings/utils/cumulativeCirculatingSupply.ts

```diff
@@ -0,0 +1,329 @@
+            totalIssuance = await totalIssuanceStorage.getAsV17(param as CurrencyId_V17);
+        } else if (totalIssuanceStorage.isV1020000) {
+            totalIssuance = await totalIssuanceStorage.getAsV1020000(param as CurrencyId_V1020000);
+        } else if (totalIssuanceStorage.isV1021000) {
+            totalIssuance = await totalIssuanceStorage.getAsV1021000(param as CurrencyId_V1021000);
+        } else {
+            ctx.log.warn("UNKOWN STORAGE VERSION: tokens.totalIssuance`");
+            // update height, to avoid more failed lookups for same height
+            totalIssuanceLatest.height = block.height;
+            // return last known value
+            return totalIssuanceLatest.value;
+        }
+    }
+
+    totalIssuanceLatest.height = block.height;
+    totalIssuanceLatest.value = totalIssuance;
+
+    return totalIssuance;
+}
+
+function findEntityBefore(
+    timestampedEntities: CumulativeCirculatingSupply[],
+    beforeTimestamp: Date,
+    customFilter?: (elem: CumulativeCirculatingSupply) => boolean
+): CumulativeCirculatingSupply | undefined {
+    if (timestampedEntities.length < 1) {
+        return undefined;
+    }
+
+    // find the latest entry that has a tillTimestamp of less than the search timestamp
+    const timestampFiltered = timestampedEntities
+        .filter(entity => entity.tillTimestamp.getTime() < beforeTimestamp.getTime());
+
+    // apply custom filter (most likely checking poolId for volume per pool entities)
+    const customFiltered = (customFilter == undefined) ? timestampFiltered : timestampFiltered.filter(customFilter);
+
+    return customFiltered.reduce((prev: CumulativeCirculatingSupply | undefined, current) => {
+        if (prev === undefined) {
+            return current;
+        }
+
+        return prev.tillTimestamp.getTime() >
+            current.tillTimestamp.getTime()
+            ? prev
+            : current;
+    },
+    undefined);
+}
+
+async function recalculateAndSetCirculatingSupply(
+    entity: CumulativeCirculatingSupply
+): Promise<CumulativeCirculatingSupply> {
+    const issuance = entity.totalIssuance;
+    const locked = entity.amountLocked;
+    const reserved = entity.amountReserved;
+    const system = entity.amountSystemAccounts;
+
+    const circulating = issuance - locked - reserved - system;
+    const circulatingHuman = await convertAmountToHuman(entity.currency, circulating);
+
+    entity.amountCirculating = circulating;
+    entity.amountCirculatingHuman = circulatingHuman;
+    return entity;
+}
+
+async function getInitialSupplyValues(
+    nativeToken: Token.KINT | Token.INTR,
+): Promise<Partial<CumulativeCirculatingSupply>> {
+    const currency = new NativeToken({token: nativeToken});
+    const baseInfo = {
+        symbol: nativeToken,
+        currency,
+    };
+    
+    if (!isMainnet()) {
+        // testnet values, just initialize all at zero, set starting issuance
+        const totalIssuance = 10000000000000000000n;
```

Comment:
Can we fetch this from the chain? It might be incorrect for testnet

Reply:
Yes, we can (and it does for subsequent events). Good shout, thank you. Let me adjust that.

Reply:
:white_check_mark: commit b62c479